psi4/driver/driver.py

```diff
@@ -873,6 +873,190 @@ def properties(*args, **kwargs):
+    core.print_out("\n  --------------------------------------------------------------------------------------------- ~")
+    core.print_out((f"\n    Convergence Criteria  {params.Convergence_energy:10.2e}    "
+                    f"{params.Convergence_gmax:10.2e}    {params.Convergence_grms:10.2e}    "
+                    f"{params.Convergence_dmax:10.2e}    {params.Convergence_drms:10.2e}    ~"))
+    core.print_out("\n  --------------------------------------------------------------------------------------------- ~")
+
+    core.print_out((f"\n   {optimizer.Iteration:4d} {optimizer.E:16.8e}    --------    "
+                    f"{gmax:10.2e} {conv_gmax}  {grms:10.2e} {conv_grms}    --------      --------    ~"))
+    while True:
+        if optimizer.state == geometric.optimize.OPT_STATE.CONVERGED:
+            core.print_out("\n\n  Optimization converged!                                                                       ~\n")
+            break
+        elif optimizer.state == geometric.optimize.OPT_STATE.FAILED:
+            core.print_out("\n\n  Optimization failed to converge!                                                              ~\n")
+            break
+        optimizer.step()
+        optimizer.calcEnergyForce()
+        optimizer.evaluateStep()
+        grms, gmax = optimizer.calcGradNorm()
+        drms, dmax = geometric.optimize.calc_drms_dmax(optimizer.X, optimizer.Xprev)
+        conv_energy = '*' if np.abs(optimizer.E - optimizer.Eprev) < params.Convergence_energy else ' '
+        conv_gmax = '*' if gmax < params.Convergence_gmax else ' '
+        conv_grms = '*' if grms < params.Convergence_grms else ' '
+        conv_dmax = '*' if dmax < params.Convergence_dmax else ' '
+        conv_drms = '*' if drms < params.Convergence_drms else ' '
+        core.print_out((f'\n   {optimizer.Iteration:4d} {optimizer.E:16.8e}  '
+                        f'{optimizer.E-optimizer.Eprev:10.2e} {conv_energy}  {gmax:10.2e} {conv_gmax}  '
+                        f'{grms:10.2e} {conv_grms}  {dmax:10.2e} {conv_dmax}  {drms:10.2e} {conv_drms}  ~'))
+
+        if return_history:
+            step_energies.append(optimizer.E)
+            step_coordinates.append(core.Matrix.from_array(optimizer.X.reshape(-1,3)))
+            step_gradients.append(core.Matrix.from_array(optimizer.gradx.reshape(-1,3)))
+
+    return_energy = optimizer.E
+    opt_geometry = core.Matrix.from_array(optimizer.X.reshape(-1,3))
+    molecule.set_geometry(opt_geometry)
+    molecule.update_geometry()
+    core.print_out(f'\n  Final Energy : {return_energy} \n')
+    core.print_out('\n  Final Geometry : \n')
+    molecule.print_in_input_format()
+
+    if return_history:
+        history = {
+            'energy': step_energies,
+            'gradient': step_gradients,
+            'coordinates': step_coordinates,
+        }
+
+    # run gradient at optimized geometry to get a wfn
+    if return_wfn:
+        g, wfn = gradient(name, return_wfn=True, **kwargs)
```

Comment:
I got nervous about different P::e.globals or other settings depending on whether return_wfn = T/F. Even if it's fine now, may change btwn psithon/psiapi, if qcng is involved whether extras['psiapi'] = T/F, and with ddd or not. I don't particularly want to waste cycles by always running the extra gradient calc, but may be worth it to avert routing bugs.

Reply:
Are you saying we should run that gradient calculation regardless of the value of `return_wfn`? I think that would be a pretty reasonable decision. 

I wouldn't be too worried about the extra cycle. In my limited experience, GeomeTRIC usually requires many more optimization cycles than Optking, so if a user has made the decision to use GeomeTRIC, a single additional cycle would not be that much of a problem.

Reply:
Good point on number of iterations. In that case, yes, go for final gradient() call under all circumstances.

Reply:
Even better, I did away with the final gradient call. I can stick the wfn to the GeomeTRIC Engine object during the optimization, so there's no need to make any additional calls after the optimization loop.